### DISCUSSION TITLE
mapstruct generates code that must result in npe

### DISCUSSION THREAD

**Discussion by r-uu** [mapstruct/mapstruct#3170] [Q&A]
Note: I also posted a question regarding the same problem at [stackoverflow](https://stackoverflow.com/questions/75540837/lombok-with-mapstruct-mapstruct-generates-code-that-results-in-npe).

I have a simple lombok/mapstruct DTO like this:
```java
@RequiredArgsConstructor
@NoArgsConstructor(access = PROTECTED) // generate no args constructor for jsonb, jaxb, mapstruct, ...
@Getter
//@Accessors(fluent = true)
@ToString
@EqualsAndHashCode
public class DepartmentDTO
{
	@NonNull @Setter private String name;

	/** may be null to indicate that employees are not yet loaded */
	@EqualsAndHashCode.Exclude
	private Set<EmployeeDTO> employees;

	/**
	 * let this be used by mapstruct (@Default, @ObjectFactory) and make sure to manually call required args constructor
	 * @param department incoming entity to be used for construction of instance
	 * @param context incoming context to properly handling cyclic dependencies
	 */
	@Default // necessary, makes sure mapstruct does not use no-args-constructor
	public DepartmentDTO(@NonNull DepartmentEntity department, @NonNull MapStructCycleTrackingContext context)
	{
		this(department.getName());
...
```
My mapstruct mapper is implemented like this:
```java
@Mapper
public interface MapStructMapper
{
	MapStructMapper INSTANCE = Mappers.getMapper(MapStructMapper.class);

	DepartmentEntity map(DepartmentDTO    department, MapStructCycleTrackingContext context);
```
Now the mapstruct generated code looks like this:
```java
    @Override
    public DepartmentEntity map(DepartmentDTO department, MapStructCycleTrackingContext context) {
        if ( department == null && context == null ) {
            return null;
        }

        MapStructCycleTrackingContext context1 = null;
        context1 = context;

        DepartmentDTO department1 = null;

        DepartmentEntity departmentEntity = new DepartmentEntity( department1, context1 );
```
In the last two lines it is obvious that the constructor is always called with ```null``` as the first parameter value. No surprise that this leads to a NPE inside the constructor. What am I doing wrong that makes mapstruct generate this incorrect code? You can find a minimal reproducer [here](https://github.com/r-uu/sandbox.mapstruct) (package "department_employee_bidirectional" under src/test/java). Please look out for tag "mapstruct-generated-code-produces-NPE".

edit:

I created a more minimal reproducer [here](https://github.com/r-uu/sandbox.mapstruct.unidirectional).
The context parameter exists because I plan to extend the code to bidirectional relation with correct cycle handling later.

**Comment by filiphr** [comment]
Thanks for reporting this. This looks like some weird bug. The reason why this happens is because there is a setter for `name` and that makes the source parameter `department` not to be used as an option for the constructor mapping.

If I add `@Mapping(target = "department", source = "department")` or I remove the setter for `name` then it works as expected.  In any case the example is rather complex. Why do you have such a complex DTO (it has a setter for name, but it also has a different constructor for MapStruct)
